clash-lib/src/Clash/Core/Evaluator/Semantics.hs

```diff
@@ -0,0 +1,548 @@
+-- the type lambda with the current environment.
+--
+evaluateTyLam :: TyVar -> Term -> Eval Value
+evaluateTyLam i x = State.gets (VTyLam i x)
+
+-- | Default implementation for evalating a term / type application. This
+-- checks if the arguments are applied to a data constructor or primitive, and
+-- creates a value if it is fully applied. If the arguments are applied to a
+-- function, the first argument is applied.
+--
+evaluateAppWith
+  :: (Term -> Eval Value)
+  -> (PrimInfo -> [Either Value Type] -> Eval Value)
+  -> (Value -> Either Value Type -> Eval Value)
+  -> Term
+  -> Either Term Type
+  -> Eval Value
+evaluateAppWith eval evalPrim apply x y
+  | Data dc <- f
+  = if isFullyApplied (dcType dc) args
+       then VData dc <$> evalArgs args
+       else etaExpand term >>= eval
+
+  | Prim p <- f
+  , nArgs  <- length . fst $ splitFunForallTy (primType p)
+  = case compare (length args) nArgs of
+      LT -> etaExpand term >>= eval
+      EQ -> evalArgs args >>= evalPrim p
+      GT -> do
+        (pArgs, rArgs) <- splitAt nArgs <$> evalArgs args
+        primRes <- evalPrim p pArgs
+        foldM apply primRes rArgs
+
+  | otherwise
+  = do
+       evalF  <- eval f
+       evalAs <- evalArgs args
+       foldM apply evalF evalAs
+ where
+  term      = either (App x) (TyApp x) y
+  (f, args) = collectArgs term
+  evalArg   = bitraverse eval pure
+  evalArgs  = traverse evalArg
+
+-- | Default implementation for evaluating a letrec expression. This adds all
+-- bindings to the heap without eagerly evaluating them, then evaluates the
+-- body of the expression.
+--
+-- Bindings are not evaluated eagerly, as they may not be needed depending how
+-- the body of the expression evaluates (i.e. bindings may only be used in a
+-- pruned alternative of a case expression).
+--
+evaluateLetrecWith
+  :: (Term -> Eval Value)
+  -> [LetBinding]
+  -> Term
+  -> Eval Value
+evaluateLetrecWith eval bs x =
+  State.withState addTerms (eval x)
+ where
+  addTerms env =
+    let terms = fmap (fmap Left) bs
+     in foldl' (flip $ uncurry insertLocal) env terms
+
+-- | Default implementation for evaluating a case expression. This replaces the
+-- entire expression with the chosen alternative if it can be statically
+-- determined, evaluating that branch only. If there is ambiguity over which
+-- alternative to choose, all alternatives are evaluated.
+--
+-- Predicates are given for determining if a literal or data constructor is
+-- matched by a pattern. This is needed as in some frontends (e.g. GHC), there
+-- may be types that can be matched with literals or constructors. For example
+--
+--   case (x :: Integer) of
+--     1 -> ...
+--
+-- and
+--
+--   case (x :: Integer) of
+--     S# 1 -> ...
+--
+-- match the same value, but do so in different ways. This means the function
+-- to check if a literal is matched would have to check both literal patterns
+-- and data patterns for the constructors of Integer.
+--
```

Comment:
I think this needs a TODO on what to do when the subject of the case expression evaluates to bottom. That is, if the subject evaluates to bottom, then the entire case expression evaluates to bottom.